### ISSUE TITLE
Specialize for constants on the top of the stack

### ISSUE THREAD

**Issue by apt1002** [rrthomas/smite#126]
Several SMite instructions are parameterised by typically constant small value on top of the stack. These instructions admit better implementations if the constant is statically known. For example:

 - LIT SWAP
 - LIT POP
 - LIT LOAD
 - LIT STORE
 - LIT SIGN_EXTEND

The special opcodes LIT_0, LIT_1, LIT_2 and LIT3, used instead of LIT in the above examples, give the specializing interpreter the necessary information. Exploit these.

Suggested approach:

 - [x] In gen-labels, detect the small literals, and record that the following state has a constant on top of the stack.
 - [x] When generating an action routine for a variadic instruction, if the top stack item is a known constant, replace the instruction with a specialized one.

This overlaps a bit with #125. In the case where the small literal and the variadic instruction are encoded into the same word, the C compiler might be able to do the specialization in some cases. However, non-variadic specializations of POP and SWAP would handle more cases, and we should also handle the cases that are split across words.

**Comment by apt1002** [comment]
On branch "specializer" there is now code to track whether the top stack item is a constant.

Variadic instructions are not guessed unless the top stack item is a constant. This results in a slow-down of less than 1%, while eliminating a difficult case.

The generated code for a `DUP` instruction currently looks like this:
```
    A_25:
        assert(res == MIT_ERROR_OK);
        assert(cached_depth == 1);
        assert(stack_0 == 0); \\ <-------------------------------- Look!
        if (opcode == O_DUP) {
            mit_word x;
            mit_word COUNT;
            COUNT = stack_0;
            if ((S->STACK_DEPTH < (mit_uword)(2 + COUNT))) {
                S->BAD = 2 + COUNT - 1;
                RAISE(MIT_ERROR_INVALID_STACK_READ);
            }
            x = *UNCHECKED_STACK(1 + COUNT);
            S->STACK_DEPTH -= 2 + COUNT;
            cached_depth = 0;
            cached_depth = 1;
            S->STACK_DEPTH += 2 + COUNT;
            *UNCHECKED_STACK(1 + COUNT) = x; \\ <--------------------- Look!
            stack_0 = x;
            FETCH;

            goto A_41;
        }

        goto A_42;
```

Note that the value of `COUNT` is a compile-time constant. This improves the code for the overflow check and for loading `x` from the stack slot. However, the job is not done yet, because `x` is still written to the stack unnecessarily at the end.

Furthermore, in states where `cached_depth` is larger than `1` on entry, stack items are still flushed to memory unnecessarily before `x` is loaded, and `x` is loaded from memory unnecessarily.

The generated code for a `LOAD` instruction currently looks like this:
```
    A_8:
        assert(res == MIT_ERROR_OK);
        assert(cached_depth == 1);
        assert(stack_0 == 2); // <-------------------------------- Look!
        if (opcode == O_LOAD) {
            mit_word x;
            mit_word size;
            mit_word addr;
            if ((S->STACK_DEPTH < (mit_uword)(2))) {
                S->BAD = 2 - 1;
                RAISE(MIT_ERROR_INVALID_STACK_READ);
            }
            addr = *UNCHECKED_STACK(1);
            size = stack_0;
            S->STACK_DEPTH -= 2;
            cached_depth = 0;
            int ret = load(S, addr, size, &x); // <------------------------- Look!
            if (ret != 0)
                RAISE(ret);
            cached_depth = 1;
            S->STACK_DEPTH += 1;
            stack_0 = x;
            FETCH;

            goto A_16;
        }

        goto A_17;
```

Note that the value of `size` is a compile-time constant. Unfortunately, this appears to be insufficient to persuade GCC to inline the call to `load()`.

In summary, it still appears to be necessary to substitute specialized versions of these instructions.

**Comment by rrthomas** [comment]
Compiler-training is frustrating! In the first case, does gcc elide the unnecessary write?

**Comment by apt1002** [comment]
No.
